packages/backend/src/helpers/actions/handle-failed-step-and-throw.ts

```diff
@@ -0,0 +1,193 @@
+ * or null on failure (badly formatted value, or retry-after in the past)
+ */
+function parseRetryAfterToMs(
+  rawHeaderValue: string | null | undefined,
+): number | null {
+  if (!rawHeaderValue) {
+    return null
+  }
+
+  // Try parsing as seconds.
+  let retryAfter = Number(rawHeaderValue)
+  if (!isNaN(retryAfter)) {
+    return retryAfter >= 0 ? retryAfter * 1000 : null
+  }
+
+  // Try parsing as date.
+  retryAfter = new Date(rawHeaderValue).getTime()
+  if (isNaN(retryAfter)) {
+    return null
+  }
+
+  retryAfter = retryAfter - Date.now()
+  return retryAfter >= 0 ? retryAfter : null
+}
+
+const RETRY_AFTER_LIMIT_MS = 12 * 60 * 60 * 1000 // 12 hours
+const RETRIABLE_ERROR_SUBSTRINGS = ['ETIMEDOUT', 'ECONNRESET']
+const RETRIABLE_STATUS_CODES = [
+  504,
+  429, // Some 429s may not have Retry-After
+]
+
+function handleHttpError(
+  executionError: HttpError,
+  errorDetails: HandleFailedStepAndThrowParams['errorDetails'],
+): never {
+  // Handle reasonable Retry-After responses.
+  const retryAfterMs = parseRetryAfterToMs(
+    executionError.response?.headers?.['retry-after'],
+  )
+
+  if (retryAfterMs !== null) {
+    if (retryAfterMs <= RETRY_AFTER_LIMIT_MS) {
+      throw new RetriableError({
+        error: errorDetails,
+        delayInMs: retryAfterMs,
+        delayType: 'step',
+      })
+    }
+    throw new UnrecoverableError(`Retry-After (${retryAfterMs}ms) is too long!`)
+  }
+
+  // Handle retriable status codes
+  if (RETRIABLE_STATUS_CODES.includes(executionError.response?.status)) {
+    throw new RetriableError({
+      error: errorDetails,
+      delayInMs: 'default',
+      delayType: 'step',
+    })
+  }
+
+  // Handle retriable errors (identified by message substring)
+  const errorString = JSON.stringify(get(errorDetails, 'details.error', ''))
+  for (const errorSubstring of RETRIABLE_ERROR_SUBSTRINGS) {
+    if (errorString.includes(errorSubstring)) {
+      throw new RetriableError({
+        error: errorDetails,
+        delayInMs: 'default',
+        delayType: 'step',
+      })
+    }
+  }
+
+  // All other HTTP errors are un-retriable
+  throw new UnrecoverableError(JSON.stringify(errorDetails))
+}
+
+interface HandleFailedStepAndThrowParams {
+  executionId: Execution['id']
+  errorDetails: ExecutionStep['errorDetails']
+
+  // This is thrown from app.run, which _in theory_ can be anything.
+  executionError: unknown
+
+  context: {
+    isQueuePausable: boolean
+    span: Span
+    worker: WorkerPro<IActionJobData>
+    job: JobPro<IActionJobData>
+  }
+}
+
+export async function handleFailedStepAndThrow(
+  params: HandleFailedStepAndThrowParams,
+): Promise<never> {
+  let { executionError } = params
+  const { executionId, errorDetails, context } = params
+  const { span, job } = context
+
+  // Edge case as some actions throw StepError now, but others don't.
+  if (executionError instanceof StepError) {
+    executionError = executionError.cause
+  }
```

Comment:
just confirming, wouldn't a `StepError` that is not wrapping a `HttpError` not contain any error for you to call `.cause` on it?

Reply:
Correct. In this case, we will throw `UnrecoverableError`, which is OK.